### DISCUSSION TITLE
Chrome Canary支持HLG了，谁能写个插件让杜比视界画质选项强行显示出来

### DISCUSSION THREAD

**Discussion by SwimmingTiger** [the1812/Bilibili-Evolved#3498] [功能建议 / Ideas]
B站的杜比视界画质大多是HLG+Profile 8.4（ https://www.bilibili.com/read/cv12752764 ）。
备注：并非全是HLG，有些是PQ+Profile 5，浏览器播放会偏色（[参考案例](#discussioncomment-3151141)）。

最近字节跳动等方面给Chrome Canary添加了HEVC和HDR（PQ/HLG）支持（ https://zhuanlan.zhihu.com/p/541082191 ），所以通过启动参数`--enable-features=PlatformHEVCDecoderSupport`外加改 User-Agent 为 Mac Safari 的方法已经能在 Windows 版 Chrome Canary 里正常播放B站HDR10（PQ）节目了，不需要安装HEVC视频扩展。

但是杜比视界还不行，因为改了User-Agent也无法显示这个画质选项，也就没机会尝试播放了。

我觉得既然B站杜比视界的基本层是HLG，Chrome Canary已经支持HLG，那么强行播放应该可以得到HDR显示效果，虽然附加的杜比视界元数据不会被应用，但基本层的HLG毕竟也是静态HDR标准，看HLG总比看SDR要好。

谁有兴趣做个插件把杜比视界画质显示出来？我可以帮忙测试。

参考视频：https://www.bilibili.com/video/BV1Av4y1M7zN

目前只能靠MPV输出插件看HDR，非常不方便。

![image](https://user-images.githubusercontent.com/4986069/179055076-11027612-38a1-4526-b762-c9344a658632.png)

![image](https://user-images.githubusercontent.com/4986069/179055224-557dba43-ee67-46f3-9d38-2d1bc1c74125.png)

下载下来的杜比视界视频放进`video`标签里用`Chrome Canary --enable-features=PlatformHEVCDecoderSupport`观看，可以呈现HDR效果（截图过曝，因为截图工具不支持HDR），所以这个方案应该是可行的。

![image](https://user-images.githubusercontent.com/4986069/179056483-2f18dcef-5d6d-4f20-b7f8-7abc58e95f86.png)

**Comment by SwimmingTiger** [comment]
概念验证：

把视频下载插件输出的下载地址替换到B站播放器的`<video src>`里，结果可以正常播放，并且呈现了HDR效果，只是没声音（因为音频是单独的文件）。

![image](https://user-images.githubusercontent.com/4986069/179058752-91ca36e3-542d-4acb-84d3-d14f2958ce8c.png)

所以只要绕过杜比支持检查，让杜比视界画质开关强行显示出来，应该也能正常播放。

**Comment by SwimmingTiger** [comment]
### 手动执行的概念验证代码

虽然找不到怎么绕过杜比视界支持检查，但 Bilibili-Evolved 的下载器插件可以提供杜比视界视频地址，
所以可以创建一个新的video标签来播放这个视频地址，然后把原始视频隐藏并仅播放其音频。

这样就能粗略实现Windows上B站杜比视界视频的浏览器内HDR播放。

1. Windows设置里打开HDR。

   ![image](https://user-images.githubusercontent.com/4986069/179088345-89f62649-ca8a-43bb-988f-bcd87d97061f.png)

2. 安装 [Chrome Canary](https://www.google.com/intl/zh-CN/chrome/canary/)，在其快捷方式末尾添加以下参数并启动。

   ```
    --enable-features=PlatformHEVCDecoderSupport
   ```

   ![image](https://user-images.githubusercontent.com/4986069/179086782-b0c5c345-eb9f-48ca-ae6a-e56bd2d57ed8.png)

3. 打开杜比视界视频，比如 https://www.bilibili.com/video/BV1Av4y1M7zN

4. 把画质切换为360P，默认画质只是为了提供音频，所以视频清晰度越低越好，高了容易卡顿，导致音画不同步。

   杜比视界的音频文件播放不了，浏览器不支持，所以只能播放原本视频的音频。

5. 安装 Bilibili-Evolved 及**下载器**插件。

   ![image](https://user-images.githubusercontent.com/4986069/179087892-0ce895cf-1136-43ed-8b7a-98278dd56a20.png)

6. 使用**下载视频**功能下载杜比视界视频文件，输出到Toast：

   ![image](https://user-images.githubusercontent.com/4986069/179068734-3f146b4a-27be-4b65-87df-639affe88e48.png)

7. 保持Toast上的视频下载链接不关闭，然后打开F12开发者控制台，执行如下代码：

```
var video = document.querySelector('video');
video.pause();

var hdrVideo = video.cloneNode();
hdrVideo.src = document.querySelectorAll('.toast-card-message a')[0].href;
hdrVideo.currentTime = video.currentTime;

video.style.display = 'none';
video.parentElement.appendChild(hdrVideo);

// 用于两个video播放进度同步的变量
const maxDiff = 0.2; // 同步延迟容限，如果很卡可以改大
var synced = true;
var syncCount = 0;
var unsyncCount = 0;
var muted = video.muted;

// 播放暂停同步
video.addEventListener('pause', () => hdrVideo.isWaiting || hdrVideo.pause());
video.addEventListener('play', () => hdrVideo.play());
// 拖进度同步
var seeking = false;
video.addEventListener('seeking', () => {
    if (!synced) {
        return;
    }
    seeking = true;
    hdrVideo.currentTime = video.currentTime;
});
video.addEventListener('seeked', () => {
    if (!synced) {
        return;
    }
    seeking = false;
});

// 检测音画不同步并自动再次同步
video.addEventListener('timeupdate', () => {
    var diff = video.currentTime - hdrVideo.currentTime;
    console.log(
        video.currentTime,
        hdrVideo.currentTime,
        video.currentTime - hdrVideo.currentTime,
        synced,
        syncCount,
        unsyncCount,
        muted,
        video.muted
    );
    if (seeking) {
        return;
    }
    if (Math.abs(diff) > maxDiff) {
        syncCount = 0;
        unsyncCount++;
        if (unsyncCount >= 5) {
            unsyncCount = 0;
            // 不同步时自动静音
            if (synced) {
                synced = false;
                muted = video.muted;
                video.muted = true;
            }
            if (diff > 0) {
                video.currentTime = hdrVideo.currentTime;
            } else {
                hdrVideo.currentTime = video.currentTime;
            }
        }
    } else {
        syncCount++;
        unsyncCount = 0;
        if (!synced) {
            if (syncCount >= 5) {
                // 恢复同步，解除静音
                synced = true;
                video.muted = muted;
            }
        }
    }
});
```

8. 再次播放视频，即可看到杜比视界的HLG播放效果（没有动态元数据，但有基础的HDR效果）。

#### 概念验证代码的已知问题

1. 视频卡顿可能导致反复自动静音，可以尝试暂停一会儿再播放，或者往回拖进度。

2. 音画未完全同步，延迟容限是正负200ms。

3. 无法后台播放，因为后台播放会导致音画立即不同步，然后开始自动静音和反复重新同步。
   从后台返回后，视频很可能无法继续播放。

4. 视频的缓冲进度不再可见，进度条显示的只是音频的缓冲进度。

5. 只能播放当前视频，切换视频后会黑屏。

**Comment by lns103** [comment]
这个目前可能会有一个问题，b站杜比视界有profile 5的，比如[BV1dK4y1M7cY](https://www.bilibili.com/video/BV1dK4y1M7cY) ，但似乎没有做出区别，这可能需要b站官方在视频资源JSON里做出区别才能解决

**Reply by SwimmingTiger** [reply]
可以播放，只是偏色了。如果程序不能区分，那就让用户自己区分嘛。如果杜比视界可以轻松的开启关闭，看一眼不对再关掉也没什么损失。

![image](https://user-images.githubusercontent.com/4986069/179178356-a46147ca-71e1-4f11-896e-36b072c10876.png)
